File: core/core-domain/src/main/java/com/nmh/commerce/coupon/Coupon.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.domain.Money;
import com.nmh.commerce.product.Product;
import lombok.Builder;
import lombok.Getter;

import java.util.List;

@Getter
public class Coupon {
    private final Long id;
    private final String name;
    private final List<DiscountPolicy> discountPolicies;
    private final List<ConstraintPolicy> constraintPolicies;
    private final List<ExpirationPeriodPolicy> expirationPeriodPolicies;

    @Builder
    private Coupon(Long id, String name, List<DiscountPolicy> discountPolicies, List<ConstraintPolicy> constraintPolicies, List<ExpirationPeriodPolicy> expirationPeriodPolicies) {
        this.id = id;
        this.name = name;
        this.discountPolicies = discountPolicies;
        this.constraintPolicies = constraintPolicies;
        this.expirationPeriodPolicies = expirationPeriodPolicies;
    }

    public Money apply(Product product) {
        constraintPolicies.forEach(constraintPolicy -> constraintPolicy.verify(product));
        return discountPolicies.stream()
            .map(discountPolicy -> discountPolicy.calculateDiscount(product.getPrice()))
            .reduce(Money.ZERO, Money::plus);
    }
}

File: core/core-domain/src/main/kotlin/com/nmh/commerce/coupon/Coupon.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.domain.Money
import com.nmh.commerce.product.Product
import com.nmh.commerce.user.User
import java.time.LocalDateTime
import java.util.function.Consumer

data class Coupon(
    val id: Long,
    val name: String,
    val discountPolicies: MutableList<DiscountPolicy>,
    val constraintPolicies: MutableList<ConstraintPolicy>,
    val expirationPeriodPolicies: MutableList<ExpirationPeriodPolicy>,
) {
    fun apply(product: Product): Money {
        constraintPolicies.forEach(
            Consumer { constraintPolicy: ConstraintPolicy -> constraintPolicy.verify(product) },
        )
        return discountPolicies
            .stream()
            .map { discountPolicy: DiscountPolicy -> discountPolicy.calculateDiscount(product.price) }
            .reduce(Money.ZERO) { obj: Money, other: Money -> obj.plus(other) }
    }

    fun issue(user: User, issuedAt: LocalDateTime): IssuedCoupon = IssuedCoupon.issue(
        coupon = this,
        user = user,
        issuedAt = issuedAt,
    )
}
